Makefile

```diff
@@ -1,8 +1,100 @@
 dev_build_version=$(shell git describe --tags --always --dirty)
 
+.PHONY: ci-all
+ci-all: ci-bigtable ci-storage
+
+.PHONY: ci-bigtable
+ci-bigtable: bigtable
+	SERVICE=bigtable make deps
+	SERVICE=bigtable make checkgofmt
+	# SERVICE=bigtable make vet
+	SERVICE=bigtable make staticcheck
+	SERVICE=bigtable make ineffassign
+	SERVICE=bigtable make predeclared
+	SERVICE=bigtable make test
+
+.PHONY: ci-storage
+ci-storage: storage
+	SERVICE=storage make deps
+	SERVICE=storage make checkgofmt
+	SERVICE=storage make vet
+	SERVICE=storage make staticcheck
+	SERVICE=storage make ineffassign
+	SERVICE=storage make predeclared
+	SERVICE=storage make test
+
+.PHONY: bigtable
+bigtable:
+	$(eval SERVICE := bigtable)
+
+.PHONY: storage
+storage:
+	$(eval SERVICE := bigtable)
+
+.PHONY: deps
+deps:
+	cd ${SERVICE} && go get -d -v -t ./...
+
+.PHONY: updatedeps
+updatedeps:
+	cd ${SERVICE} && go get -d -v -t -u -f ./...
+
+.PHONY: install
+install:
+	cd ${SERVICE} && go install -ldflags '-X "main.version=dev build $(dev_build_version)"' ./...
+
+.PHONY: release
+release:
+	@GO111MODULE=on go install github.com/goreleaser/goreleaser
+	goreleaser --rm-dist
+
 .PHONY: docker
 docker:
 	@echo $(dev_build_version) > VERSION
-	docker build --target=gcsemulator -t fullstorydev/gcsemulator:$(dev_build_version) -t fullstorydev/gcsemulator:latest .
-	docker build --target=cbtemulator -t fullstorydev/cbtemulator:$(dev_build_version) -t fullstorydev/cbtemulator:latest .
+	docker build --target=gcsemulator -t fullstorydev/gcsemulator:$(dev_build_version)
+	docker build --target=cbtemulator -t fullstorydev/cbtemulator:$(dev_build_version)
 	@rm VERSION
+
+.PHONY: checkgofmt
+checkgofmt:
+	cd ${SERVICE} && gofmt -s -l .
+	@if [ -n "$$(cd ${SERVICE} && gofmt -s -l .)" ]; then \
+		exit 1; \
+	fi
+
+.PHONY: vet
+vet:
+	cd ${SERVICE} && go vet ./...
+
+# This all works fine with Go modules, but without modules,
+# CI is just getting latest master for dependencies like grpc.
+.PHONY: staticcheck
+staticcheck:
+	@GO111MODULE=on go install honnef.co/go/tools/cmd/staticcheck
+	cd ${SERVICE} && staticcheck ./...
+
+.PHONY: ineffassign
+ineffassign:
+	@GO111MODULE=on go install github.com/gordonklaus/ineffassign
+	cd ${SERVICE} && ineffassign .
+
+.PHONY: predeclared
+predeclared:
+	@GO111MODULE=on go install github.com/nishanths/predeclared
+	cd ${SERVICE} && predeclared ./...
+
+# Intentionally omitted from CI, but target here for ad-hoc reports.
+.PHONY: golint
+golint:
+	@GO111MODULE=on go install golang.org/x/lint/golint
+	cd ${SERVICE} && golint -min_confidence 0.9 -set_exit_status ./...
+
+# Intentionally omitted from CI, but target here for ad-hoc reports.
+.PHONY: errcheck
+errcheck:
+	@GO111MODULE=on go install github.com/kisielk/errcheck
+	cd ${SERVICE} && errcheck ./...
+
+.PHONY: test
+test:
+	cd ${SERVICE} && go test -race ./...
```

Comment:
Hmm this seems a bit clunky in general.  Is this actually better than just having separate makefiles?  Or maybe there's some more generic way of doing this?

Reply:
I think if we put it in the inner folder, most of the makefile will be repeated and we wont be reusing anything. Having it in the root, we have everything on the same place, we reuse all targets, its easier to add a new module, and we can run everything from the root

I agree with you that is clunky, but I didnt have any better idea :/

Reply:
Maybe @jhump has a better idea?  I'm not a makefile wizard.  But this seems strictly worse than just having slightly repetitive Makefiles per product.  At least the way it is, while not DRY, is clear.  This is confusing as hell.  And most of the visible top level targets simply don't work at all.

![Screen Shot 2021-08-24 at 1 48 00 PM](https://user-images.githubusercontent.com/1919660/130665035-fddc5a4d-b9b6-4229-850e-683156a3ec2a.png)

Reply:
@dragonsinth, I did talk to @franklinlindemberg a bit about alternatives.

This seemed better to me than the significant boiler-plate to have each target do the same thing N times (N = 2 for now). And @franklinlindemberg didn't care for copying the makefiles since it's maintenance overhead (every change requires remembering to touch N files), which I can understand. Also, having separate makefiles implies moving the `go.mod` contents down a level, into each of the other modules, which means a little more repetition and more ways for the sub-modules to get out of sync inadvertently... As far as the weird error you got, that could be remedied with adding [a guard](https://stackoverflow.com/questions/4728810/how-to-ensure-makefile-variable-is-set-as-a-prerequisite#7367903) as a dependency of each target that relies on that var.

Here's another idea. See what you think:
1. Have a top-level `Makefile` whose only target is `ci`, and it recursively re-invokes make for each sub-module: `cd bigtable && make ci`
2. Have another top-level file named `Makefile.shared` which contains all of the other targets and assumes relative to current directory (no `cd` statements).
3. Create symlinks for each sub-module, e.g. `bigtable/Makefile` --> `../Makefile.shared`

This could help to at least make the Makefiles themselves more DRY (though it still means moving the tool dependencies out of a top-level `go.mod` and into each of the sub-module `go.mod` files).

All in all, I don't particularly have a strong opinion on how to organize these. There are significant drawbacks/aesthetic issues with all of these approaches. 🤷 

TBH, it would probably be simplest to just make this a single-module repo. The value of each emulator in a distinct Go module is unclear and possibly very low; and it clearly makes management of repo-wide tasks a pain. Perhaps it would be better to switch to multiple modules later, after [Go workspace mode](https://go.googlesource.com/proposal/+/master/design/45713-workspace.md) is introduced? It's raison d'être is to make multi-module repos easier to work with (so it would benefit `mn`, too). Though, admittedly, I don't know what the timeline is for that...

Reply:
I got your problem, the way I did was mainly to always run ci-all. 

It didnt work because you didnt set the SERVICE env, you would need to run `SERVICE=bigtable make test`.

Another option is to create targets like test-all.

Either way, Im not sure we have a perfect solution and I would be fine putting them back in the folder, just wanted to brainstorm a little bit about this. lets see what Josh thinks

Reply:
nit: if you stick with this kind of approach (env var controlling which module to work with), I think a better name for the variable is `MODULE` instead of `SERVICE`. Just 2 cents...

Reply:
> 1. Have a top-level `Makefile` whose only target is `ci`, and it recursively re-invokes make for each sub-module: `cd bigtable && make ci`
> 2. Have another top-level file named `Makefile.shared` which contains all of the other targets and assumes relative to current directory (no `cd` statements).
> 3. Create symlinks for each sub-module, e.g. `bigtable/Makefile` --> `../Makefile.shared`

This seems fine if you guys are insistent on DRYing everything up.  (I personally don't have a problem with the Makefiles the way they are.)

I don't see much maintenance burden because I don't see that the Makefiles are going to change much.

I am pretty adamant about keeping the modules separated (just like `cloud.google.com/go/*` itself) because I hate dependency spirals and forcing users to get more deps than they actually need.